pkg/controller/restic_repository_controller.go

```diff
@@ -0,0 +1,281 @@
+	if repoManager.CheckRepo(name) == nil {
+		return nil
+	}
+
+	return repoManager.InitRepo(name)
+}
+
+func (c *resticRepositoryController) runMaintenanceIfDue(req *v1.ResticRepository, log logrus.FieldLogger) error {
+	log.Debug("resticRepositoryController.runMaintenanceIfDue")
+
+	now := c.clock.Now()
+
+	if !dueForMaintenance(req, now) {
+		log.Debug("not due for maintenance")
+		return nil
+	}
+
+	log.Info("Running maintenance on restic repository")
+
+	log.Debug("Checking repo before prune")
+	if err := c.resticMgr.CheckRepo(req.Name); err != nil {
+		return c.patchResticRepository(req, repoNotReady(err.Error()))
+	}
+
+	// prune failures should be displayed in the `.status.message` field but
+	// should not cause the repo to move to `NotReady`.
+	log.Debug("Pruning repo")
+	if err := c.resticMgr.PruneRepo(req.Name); err != nil {
+		log.WithError(err).Warn("error pruning repository")
+		if patchErr := c.patchResticRepository(req, func(r *v1.ResticRepository) {
+			r.Status.Message = err.Error()
+		}); patchErr != nil {
+			return patchErr
+		}
+	}
+
+	log.Debug("Checking repo after prune")
+	if err := c.resticMgr.CheckRepo(req.Name); err != nil {
+		return c.patchResticRepository(req, repoNotReady(err.Error()))
+	}
+
+	return c.patchResticRepository(req, func(req *v1.ResticRepository) {
+		req.Status.LastMaintenanceTime = metav1.Time{Time: now}
+	})
+}
+
+func dueForMaintenance(req *v1.ResticRepository, now time.Time) bool {
+	return req.Status.LastMaintenanceTime.Add(req.Spec.MaintenanceFrequency.Duration).Before(now)
+}
+
+func (c *resticRepositoryController) checkNotReadyRepo(req *v1.ResticRepository, log logrus.FieldLogger) error {
+	log.Info("Checking restic repository for readiness")
+
+	// we need to ensure it (first check, if check fails, attempt to init)
+	// because we don't know if it's been successfully initialized yet.
+	if err := ensureRepo(req.Name, c.resticMgr); err != nil {
+		return c.patchResticRepository(req, repoNotReady(err.Error()))
+	}
+
+	return c.patchResticRepository(req, repoReady())
+}
+
+func repoNotReady(msg string) func(*v1.ResticRepository) {
+	return func(r *v1.ResticRepository) {
+		r.Status.Phase = v1.ResticRepositoryPhaseNotReady
+		r.Status.Message = msg
+	}
+}
+
+func repoReady() func(*v1.ResticRepository) {
+	return func(r *v1.ResticRepository) {
+		r.Status.Phase = v1.ResticRepositoryPhaseReady
+		r.Status.Message = ""
+	}
+}
+
+// patchResticRepository mutates req with the provided mutate function, and patches it
+// through the Kube API. After executing this function, req will be updated with both
+// the mutation and the results of the Patch() API call.
+func (c *resticRepositoryController) patchResticRepository(req *v1.ResticRepository, mutate func(*v1.ResticRepository)) error {
+	// Record original json
+	oldData, err := json.Marshal(req)
+	if err != nil {
+		return errors.Wrap(err, "error marshalling original ResticRepository")
+	}
+
+	// Mutate
```

Comment:
This is nitpicky, but is the comment necessary here, given it's exactly the same as the function being called?